sklearn/metrics/_dist_metrics.pyx.tp

```diff
@@ -424,6 +570,86 @@ cdef class DistanceMetric{{name_suffix}}:
         """
         return dist
 
+    def _pairwise_dense_dense(self, X, Y):
+        cdef cnp.ndarray[{{INPUT_DTYPE_t}}, ndim=2, mode='c'] Xarr
+        cdef cnp.ndarray[{{INPUT_DTYPE_t}}, ndim=2, mode='c'] Yarr
+        cdef cnp.ndarray[DTYPE_t, ndim=2, mode='c'] Darr
+
+        Xarr = np.asarray(X, dtype={{INPUT_DTYPE}}, order='C')
+        self._validate_data(Xarr)
+        if X is Y:
+            Darr = np.empty((Xarr.shape[0], Xarr.shape[0]),
+                            dtype=DTYPE, order='C')
+            self.pdist(Xarr, Darr)
+        else:
+            Yarr = np.asarray(Y, dtype={{INPUT_DTYPE}}, order='C')
+            self._validate_data(Yarr)
+            Darr = np.empty((Xarr.shape[0], Yarr.shape[0]),
+                            dtype=DTYPE, order='C')
+            self.cdist(Xarr, Yarr, Darr)
+        return Darr
+
+    def _pairwise_sparse_sparse(self, X, Y):
+        X_csr = X.tocsr()
+        n_X, size = X_csr.shape
+        X_data = np.asarray(X_csr.data, dtype={{INPUT_DTYPE}})
+        X_indices = np.asarray(X_csr.indices, dtype=SPARSE_INDEX_TYPE)
+        X_indptr = np.asarray(X_csr.indptr, dtype=SPARSE_INDEX_TYPE)
+
+        if X is Y:
+            Darr = np.empty((n_X, n_X), dtype=DTYPE, order='C')
+            self.csr_pdist(
+                x1_data=X_data,
+                x1_indices=X_indices,
+                x1_indptr=X_indptr,
+                size=size,
+                D=Darr,
+            )
+        else:
+            Y_csr = Y.tocsr()
+            n_Y, _ = Y_csr.shape
+            Y_data = np.asarray(Y_csr.data, dtype={{INPUT_DTYPE}})
+            Y_indices = np.asarray(Y_csr.indices, dtype=SPARSE_INDEX_TYPE)
+            Y_indptr = np.asarray(Y_csr.indptr, dtype=SPARSE_INDEX_TYPE)
+
+            Darr = np.empty((n_X, n_Y), dtype=DTYPE, order='C')
+            self.csr_cdist(
+                x1_data=X_data,
+                x1_indices=X_indices,
+                x1_indptr=X_indptr,
+                x2_data=Y_data,
+                x2_indices=Y_indices,
+                x2_indptr=Y_indptr,
+                size=size,
+                D=Darr,
+            )
+        return Darr
+
+    def _pairwise_sparse_dense(self, X, Y):
+        X_csr = X.tocsr()
+        n_X, size = X_csr.shape
+        X_data = np.asarray(X_csr.data, dtype={{INPUT_DTYPE}})
+        X_indices = np.asarray(X_csr.indices, dtype=SPARSE_INDEX_TYPE)
+        X_indptr = np.asarray(X_csr.indptr, dtype=SPARSE_INDEX_TYPE)
+
```

Comment:
```suggestion

        # To avoid introducing redundant implementations for the CSR x dense array
        # case, we wrap the dense array into a fake CSR datastructure and leverage
        # the existing code for the CSR x CSR case.
        # The true CSR representation of a dense array would require allocating
        # a Y_indices matrix of shape (n_samples, n_features) with repeated
        # contiguous integers from 0 to n_features - 1 on each row which would
        # be very wasteful from a memory point of view. Instead we only allocate
        # a single row and adapt the CSR x CSR routines to use a modulo operation
        # when accessing Y_indices in order to achieve the same result with having
        # to materialize the indices repetition explicitly.
```